src/ast/values.ts

```diff
@@ -12,3 +32,229 @@ export const UNKNOWN_EXPRESSION: ExpressionEntity = {
 	someReturnExpressionWhenCalledAtPath: () => true,
 	toString: () => '[[UNKNOWN]]'
 };
+const returnsUnknown: RawMemberDescription = { value: { returns: UNKNOWN_EXPRESSION, callsArgs: null } };
+const callsArgReturnsUnknown: RawMemberDescription = { value: { returns: UNKNOWN_EXPRESSION, callsArgs: [0] } };
+
+export const UNKNOWN_ARRAY_EXPRESSION: ExpressionEntity = {
+	reassignPath: () => {},
+	forEachReturnExpressionWhenCalledAtPath: () => {},
+	getValue: () => UNKNOWN_VALUE,
+	hasEffectsWhenAccessedAtPath: path => path.length > 1,
+	hasEffectsWhenAssignedAtPath: path => path.length > 1,
+	hasEffectsWhenCalledAtPath: (path, callOptions, options) => {
+		if (path.length === 1) {
+			return hasMemberEffectWhenCalled(arrayMembers, path[0], callOptions, options);
+		}
+		return true;
+	},
+	someReturnExpressionWhenCalledAtPath: (
+		path: ObjectPath,
+		callOptions: CallOptions,
+		predicateFunction: SomeReturnExpressionCallback,
+		options: ExecutionPathOptions
+	) => {
+		if (path.length === 1) {
+			return someMemberReturnExpressionWhenCalled(arrayMembers, path[0], callOptions, predicateFunction, options);
+		}
+		return true;
+	},
+	toString: () => '[[UNKNOWN ARRAY]]'
+};
+const returnsArray: RawMemberDescription = { value: { returns: UNKNOWN_ARRAY_EXPRESSION, callsArgs: null } };
+const callsArgReturnsArray: RawMemberDescription = { value: { returns: UNKNOWN_ARRAY_EXPRESSION, callsArgs: [0] } };
+
+function createUnknownLiteral (type: 'boolean' | 'number' | 'string'): ExpressionEntity {
+	return {
+		reassignPath: () => {},
+		forEachReturnExpressionWhenCalledAtPath: () => {},
+		getValue: () => UNKNOWN_VALUE,
+		hasEffectsWhenAccessedAtPath: path => path.length > 1,
+		hasEffectsWhenAssignedAtPath: path => path.length > 0,
+		hasEffectsWhenCalledAtPath: path => {
+			if (path.length === 1) {
+				const subPath = path[0];
+				return isUnknownKey(subPath) || !literalMembers[type][subPath];
```

Comment:
`literalMembers[type]` can be hoisted out of this function.

Reply:
I was hoping so, too. Unfortunately there is some circular dependency between the data structures in this file (which is the main reason I moved everything into the same file). I.e., `createUnknownLiteral` is called the first time before we hit the definition of `literalMembers`. If you have a suggestion how to untangle these that would be cool but I fear it will need to remain like this for now.

Reply:
It sounds like the simplest fix may just be to flatten `UNKNOWN_LITERALS` into `UNKNOWN_LITERAL_NUMBER`, `UNKNOWN_LITERAL_BOOLEAN` and `UNKNOWN_LITERAL_STRING`. The collection object could still exist fine if lookups are needed in other places.